driver-core/src/test/unit/com/mongodb/internal/connection/IdTrackingBsonWriterSpecification.groovy

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.connection
+
+import org.bson.BsonBinaryReader
+import org.bson.BsonBinaryWriter
+import org.bson.BsonDocument
+import org.bson.RawBsonDocument
+import org.bson.codecs.BsonDocumentCodec
+import org.bson.codecs.DecoderContext
+import org.bson.codecs.EncoderContext
+import org.bson.io.BasicOutputBuffer
+import org.bson.io.BsonOutput
+import spock.lang.Specification
+
+import static org.bson.BsonHelper.documentWithValuesOfEveryType
+import static org.bson.BsonHelper.getBsonValues
+
+class IdTrackingBsonWriterSpecification extends Specification {
+
+    def 'should write all types'() {
```

Comment:
This just ensures the BsonWriter works as expected and adds an _id value if one is not set